src/Microsoft.AspNetCore.Razor/Compilation/TagHelpers/TagHelperDescriptorProvider.cs

```diff
@@ -95,39 +96,106 @@ public TagHelperDescriptorProvider(IEnumerable<TagHelperDescriptor> descriptors)
+                        IEnumerable<KeyValuePair<string, string>> matchingAttributes;
+
                         // '*' at the end of a required attribute indicates: apply to attributes prefixed with the
                         // required attribute value.
-                        if (requiredAttribute.EndsWith(
+                        if (attributeName.EndsWith(
                             RequiredAttributeWildcardSuffix,
                             StringComparison.OrdinalIgnoreCase))
                         {
-                            var prefix = requiredAttribute.Substring(0, requiredAttribute.Length - 1);
-
-                            if (!attributeNames.Any(
-                                attributeName =>
-                                    attributeName.StartsWith(prefix, StringComparison.OrdinalIgnoreCase) &&
-                                    !string.Equals(attributeName, prefix, StringComparison.OrdinalIgnoreCase)))
-                            {
-                                return false;
-                            }
+                            var prefix = attributeName.Substring(0, attributeName.Length - RequiredAttributeWildcardSuffix.Length);
+                            matchingAttributes = attributes.Where(
+                                attribute => attribute.Key.StartsWith(prefix, StringComparison.OrdinalIgnoreCase) &&
+                                    !string.Equals(attribute.Key, prefix, StringComparison.OrdinalIgnoreCase));
                         }
-                        else if (!attributeNames.Contains(requiredAttribute, StringComparer.OrdinalIgnoreCase))
+                        else
+                        {
+                            matchingAttributes = attributes.Where(
+                                attribute => string.Equals(attribute.Key, attributeName, StringComparison.OrdinalIgnoreCase));
+                        }
+
+                        if (!matchingAttributes.Any())
                         {
                             return false;
                         }
+
+                        // At least one attribute where a name matches, now validate the value (if restricted).
+                        var equalsIndex = requiredAttribute.IndexOf('=');
+                        if (equalsIndex != -1)
+                        {
+                            var afterEquals = requiredAttribute[equalsIndex + 1];
+                            var valueOffset = 0;
+                            if (afterEquals == '\'' || afterEquals == '"')
+                            {
+                                valueOffset = 1;
+                            }
+                            var valueStart = equalsIndex + 1 + valueOffset;
+                            var valueLength = requiredAttribute.Length - equalsIndex - valueOffset * 2 - 2;
+                            var requiredValue = requiredAttribute.Substring(valueStart, valueLength);
```

Comment:
E.g. instead of
```c#
!matchingAttributes.Any(
    attribute => attribute.Value.StartsWith(requiredValue, StringComparison.Ordinal))
```

Use
```c#
!matchingAttributes.Any(
    attribute => string.CompareOrdinal(attribute.Value, 0, requiredAttribute, valueStart, valueLength) == 0)
```